rush-exec/src/errors.rs

```diff
@@ -1,7 +1,70 @@
+use std::io::ErrorKind;
 use std::path::PathBuf;
+use std::{fmt, io};
 
+use anyhow::anyhow;
 use thiserror::Error;
 
+/// This is a wrapper for io::Error to add more context than the default Display.
+/// It should not be used directly. Use an internal error instead.
+#[derive(Error, Debug)]
+pub struct IoError {
+    #[from]
+    source: io::Error,
+}
+
+impl fmt::Display for IoError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.source)?;
+        if let Some(e) = self.source.get_ref() {
+            if let Some(e) = e.source() {
+                write!(f, " because: {}", e)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+pub enum FileContext {
+    Reading,
+}
+
+pub enum ProcessContext {
+    WaitingForChild
+}
+
+pub trait IoContextExt {
+    type T;
+    fn file_context(self, ctx: FileContext) -> anyhow::Result<Self::T>;
+    fn process_context(self, ctx: ProcessContext) -> anyhow::Result<Self::T>;
+}
+
+impl<T> IoContextExt for Result<T, io::Error> {
+    type T = T;
+    fn file_context(self, ctx: FileContext) -> anyhow::Result<T> {
+        self.map_err(|e| {
+            match (e.kind(), ctx) {
+                (ErrorKind::NotFound, FileContext::Reading) => anyhow!("File not found"),
+                (ErrorKind::PermissionDenied, FileContext::Reading) => anyhow!("No permission to read file"),
+                // unstable: (ErrorKind::IsADirectory, FileContext::Reading) => anyhow!("Cannot read a directory"),
```

Comment:
Which part of this is unstable?

Reply:
See [this issue](https://github.com/rust-lang/rust/issues/86442) - IsADirectory is currently unstable. 

I should've deleted it since leaving comments in is bad form, but wanted to have it there just to mention it.